migrate/json_to_terraform.py

```diff
@@ -0,0 +1,340 @@
+#!/usr/bin/env python3
+"""
+Generates terraform repos, users, and roles files for an organization,
+based on exported JSON permissions.
+
+Because there is overlap in users between organizations, this
+script will not generate module blocks for users that are already
+declared by any users_*.tf scripts within the root of this repo;
+otherwise, terraform would complain about duplicate users.
+
+Usage:
+    python -m migrate.json_to_terraform ORGANIZATION [--phony]
+Or just:
+    make migrate-json-to-terraform
+
+Requirement installation:
+    make migrate-requirements  # in a virtualenv, from repo root
+
+Other requirements:
+    * Python >=3.8
+    * Expects teams and repositories to be dumped to
+     `migrate/export-{ORGANIZATION].json`.
+
+Linting:
+    make migrate-requirements-dev  # in a virtualenv, from repo root
+    make migrate-lint
+"""
+
+
+# pylint: disable=unspecified-encoding
+import json
+import os
+import sys
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Dict, List, NewType, Set
+
+from .utils import (REPO_ADMIN, REPO_MAINTAIN, REPO_READ, REPO_WRITE,
+                    RepoAccessLevel, RepoName, TeamSlug, Username,
+                    access_level_to_string)
+
+# Types for Terraform labels.
+# A "label" is that string that goes next to resource type or module name
+# in a Terraform block.
+# For example, in these blocks:
+#   resource "resourcetype" "blah1" { ... }
+#   module "blah2" { source = "./module/moduletype", ... }
+# the labels are "blah1" and "blah2".
+# Labels are used as the primary identifier for a Terraform block.
+# Per Terraform style, they should be lowercase and contain only [a-z0-9_],
+# so not dots, dashes, or spaces.
+RoleLabel = NewType("RoleLabel", str)
+RepoLabel = NewType("RepoLabel", str)
+UserLabel = NewType("UserLabel", str)
+
+
+@dataclass
+class TerraRole:
+    """
+    A role grants a set of repository permissions to a set of users.
+
+    Each of these roles will be represented as a Terraform module
+    invocation.
+    """
+
+    label: RoleLabel
+    repo_access: Dict[RepoName, RepoAccessLevel]
+    members: Set[Username]
+
+
+def main():
+    """
+    Script entrypoint
+    """
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    assert 2 <= len(sys.argv) <= 3, "Expected 1-2 args: ORGANIZATION [--phony]"
```

Comment:
Nit for future things (not worth doing for this script): `assert` isn't meant catch runtime errors, and would be skipped if running Python with optimizations on (which almost never happens in practice).

Reply:
Wow, good to know. I've been thinking of it as "more concise way of writing `if !x: raise Exception()`" for years -- whoops.